### DISCUSSION TITLE
Better support for working with JSON data

### DISCUSSION THREAD

**Discussion by rsmmr** [zeek/zeek#267] [Ideas]
Would be nice to have better capabilities to read and write JSON data, including  
serialization/deserialization of Bro values to/from JSON. Relates to #150

**Comment by JustinAzoff** [comment]
Another subtle issue that exists now: the logging framework doesn't let you log a table, even though that would be easily supported in the json writer.

**Comment by 0xxon** [comment]
And just to add it here - the fact that the logging framework does not support tables was a design choice - it is limited to data types that all forms of output can accept.
That being said - that has not really that much to do with the dedicated json functionality that we want to add, which probably won't have anything to do with the logging framework.

**Comment by sethhall** [comment]
It may be worthwhile to revisit the idea of logging tables during this work.  It's possible we could figure out some way to make it automatically work in a reasonable way with the ascii writer too (sheesh, that writer really needs to be renamed too!)

**Comment by rsmmr** [comment]
Quick thought: we are already serializing everything to/from `broker::data`. I'm wondering if we could leverage that to produce/ingest JSON from that representation; we'd just need to a way to associate names witht the tuple elements. That would avoid having to go through all the types once more to add JSON serialization.

An alternative would be changing the existing serialization to generate dictionaries instead of broker tuples, and then go from there to both broker and JSON. Though not sure how the opposite direction would look like there.

**Reply by Neverlord** [reply]
Plus, CAF can serialize `broker::data` to/from JSON already.

**Comment by timwoj** [comment]
I talked to @sethhall about this for a bit. Here's what we came up with:

- Add a JSON input reader similar to the pcap reader
- Add a file analyzer where you could pass JSON files from other protocols to be analyzed, such as if an HTTP connection contained JSON data
- Support for JSONPointer in protocol analyzers and use the findings from them to generate events

The big issue right now is that the library we opted to use for the rest of the JSON work doesn't have any sort of streaming API. It supports input streams but that's entirely pull and not push. We might be able to work with nlohmann to get something like that added though.

**Comment by rsmmr** [comment]
@timwoj rapidjson solves that issue, right?

**Comment by timwoj** [comment]
It's been a while since I looked at this. Was the issue before that the stream API in nlohmann's didn't support something like reading part of the data and then reading more later? Rapidjson does have a streaming API (https://rapidjson.org/md_doc_stream.html) but it might have the same issue.

**Comment by rsmmr** [comment]
Yeah, I think so. Darn, yes, that looks like the Rapidjson streaming API is pull as well.
